File: src/history/format.ts
import { basename } from "node:path";
import { statusLabel } from "./project";
import type { RunDetail, RunListItem, RunProjectedStatus } from "./types";

const SEP = " · ";

function truncateGraphemes(text: string, max: number): string {
  if (max <= 0) return "";
  const chars = [...text];
  if (chars.length <= max) return text;
  if (max === 1) return "…";
  return `${chars.slice(0, max - 1).join("")}…`;
}

function formatElapsed(ms: number): string {
  const sec = Math.floor(ms / 1000);
  if (sec < 60) return `${sec}s`;
  const min = Math.floor(sec / 60);
  if (min < 60) return `${min}m${sec % 60 ? `${sec % 60}s` : ""}`;
  const hr = Math.floor(min / 60);
  return `${hr}h${min % 60 ? `${min % 60}m` : ""}`;
}

function abbreviateStatus(status: RunProjectedStatus, width: number): string {
  const full = statusLabel(status);
  if (full.length <= width) return full;
  if (status === "interrupted") return width >= 5 ? "INTR" : "I";
  if (status === "succeeded") return width >= 4 ? "OK" : "O";
  if (status === "failed") return width >= 4 ? "FAIL" : "F";
  if (status === "running") return width >= 3 ? "RUN" : "R";
  if (status === "stale") return width >= 3 ? "STL" : "S";
  return width >= 3 ? "…" : "";
}

/** One list row; priority: status, progress, elapsed, workflow, optional location. */
export function formatRunRow(
  item: RunListItem,
  width: number,
  opts: { showLocation?: boolean } = {},
): string {
  const status = abbreviateStatus(item.status, Math.min(12, Math.max(3, width)));
  const progress =
    item.progress !== undefined ? `${item.progress.done}/${item.progress.total}` : "";
  const elapsed = formatElapsed(item.elapsed_ms);
  const location = opts.showLocation ? basename(item.checkout_root) : "";
  const fixed = [status, progress, elapsed].filter(Boolean);
  const fixedWidth = fixed.reduce((n, part) => n + part.length + SEP.length, 0);
  let remain = Math.max(0, width - fixedWidth);
  let workflow = item.title || item.workflow;
  let loc = location;
  if (loc && remain > 0) {
    const locBudget = Math.min(loc.length, Math.max(0, Math.floor(remain / 3)));
    loc = truncateGraphemes(loc, locBudget);
    remain -= loc ? loc.length + SEP.length : 0;
  } else {
    loc = "";
  }
  workflow = truncateGraphemes(workflow, remain);
  return [status, workflow, progress, elapsed, loc].filter(Boolean).join(SEP);
}

export function formatRunListEmpty(opts: {
  scope: "current" | "all";
  hasMachineRuns: boolean;
  filterActive: boolean;
}): string {
  if (opts.filterActive) return "no matching runs";
  if (!opts.hasMachineRuns) return "no workflow has run yet";
  if (opts.scope === "current") return `no runs in this worktree${SEP}Ctrl+G for All`;
  return "no runs";
}

export function formatRunDetailLines(detail: RunDetail, width: number): string[] {
  if (detail.kind === "invalid") return [truncateGraphemes(detail.message ?? "invalid run", width)];
  if (detail.kind === "missing")
    return [truncateGraphemes(detail.message ?? "run not found", width)];
  if (detail.kind === "expired") return [truncateGraphemes(detail.message ?? "run expired", width)];
  if (detail.kind === "unavailable") {
    return [truncateGraphemes(detail.message ?? "history unavailable", width)];
  }
  const lines: string[] = [];
  const head = [
    statusLabel(detail.status),
    detail.title || detail.workflow,
    detail.display_id,
    formatElapsed(detail.elapsed_ms),
  ]
    .filter(Boolean)
    .join(SEP);
  lines.push(truncateGraphemes(head, width));
  lines.push(truncateGraphemes(detail.checkout_root, width));
  if (detail.status === "stale") {
    lines.push(truncateGraphemes("writer heartbeat stale — not a failure", width));
  }
  for (const step of detail.steps) {
    const indent = Math.max(0, step.workflow_path.length - 1);
    const prefix = `${"  ".repeat(indent)}${step.ordinal}/${step.total} `;
    const outcome = step.outcome ?? (step.active ? "running" : "");
    const body = `${prefix}${step.label}${outcome ? `${SEP}${outcome}` : ""}`;
    lines.push(truncateGraphemes(body, width));
    if (step.explanation) {
      lines.push(truncateGraphemes(`${"  ".repeat(indent + 1)}${step.explanation}`, width));
    }
  }
  if (detail.current_step?.active) {
    const step = detail.current_step;
    const indent = Math.max(0, step.workflow_path.length - 1);
    lines.push(
      truncateGraphemes(
        `${"  ".repeat(indent)}${step.ordinal}/${step.total} ${step.label}${SEP}running`,
        width,
      ),
    );
  }
  if (detail.remaining !== undefined && detail.remaining > 0) {
    lines.push(
      truncateGraphemes(
        `${detail.remaining} step${detail.remaining === 1 ? "" : "s"} not run`,
        width,
      ),
    );
  }
  if (detail.failure_explanation && !detail.steps.some((s) => s.explanation)) {
    lines.push(truncateGraphemes(detail.failure_explanation, width));
  }
  return lines.length > 0 ? lines : [truncateGraphemes("no step outcomes yet", width)];
}

export function formatStartingDetail(workflow: string, id: string, width: number): string[] {
  return [
    truncateGraphemes(`STARTING${SEP}${workflow}${SEP}${id.slice(0, 8)}`, width),
    truncateGraphemes("claiming run history…", width),
  ];
}

File: internal/runsbrowser/format.go
package runsbrowser

import (
	"path/filepath"
	"strconv"
	"strings"

	"github.com/aorumbayev/herdr-workflows/internal/history"
	"github.com/aorumbayev/herdr-workflows/internal/tui"
)

const detailViewport = 10

// RunListEmptyOpts names the empty-list copy inputs.
type RunListEmptyOpts struct {
	Scope          Scope
	HasMachineRuns bool
	FilterActive   bool
	Unavailable    bool
}

// FormatRunRowOpts configures list-row layout.
type FormatRunRowOpts struct {
	ShowLocation bool
}

func abbreviateStatus(status string, width int) string {
	full := history.StatusLabel(status)
	if tui.Columns(full) <= width {
		return full
	}
	switch status {
	case "interrupted":
		if width >= 5 {
			return "INTR"
		}
		return "I"
	case "succeeded":
		if width >= 4 {
			return "OK"
		}
		return "O"
	case "failed":
		if width >= 4 {
			return "FAIL"
		}
		return "F"
	case "running":
		if width >= 3 {
			return "RUN"
		}
		return "R"
	case "stale":
		if width >= 3 {
			return "STL"
		}
		return "S"
	default:
		if width >= 3 {
			return "..."
		}
		return ""
	}
}

// FormatRunRow lays out one list row: status, workflow, progress, elapsed, optional location.
func FormatRunRow(item history.Summary, width int, opts FormatRunRowOpts) string {
	status := abbreviateStatus(item.Status, min(12, max(3, width)))
	progress := ""
	if item.Progress != nil {
		progress = strconv.Itoa(item.Progress.Done) + "/" + strconv.Itoa(item.Progress.Total)
	}
	elapsed := history.FormatElapsed(item.ElapsedMs)
	location := ""
	if opts.ShowLocation {
		location = filepath.Base(item.CheckoutRoot)
	}
	fixed := []string{status}
	if progress != "" {
		fixed = append(fixed, progress)
	}
	fixed = append(fixed, elapsed)
	fixedWidth := 0
	for _, part := range fixed {
		fixedWidth += len(part) + len(tui.ChromeSep)
	}
	remain := max(0, width-fixedWidth)
	workflow := item.Title
	if workflow == "" {
		workflow = item.Workflow
	}
	loc := location
	if loc != "" && remain > 0 {
		locBudget := min(tui.Columns(loc), max(0, remain/3))
		loc = tui.Truncate(loc, locBudget)
		if loc != "" {
			remain -= tui.Columns(loc) + tui.Columns(tui.ChromeSep)
		}
	} else {
		loc = ""
	}
	workflow = tui.Truncate(workflow, remain)
	parts := []string{status, workflow}
	if progress != "" {
		parts = append(parts, progress)
	}
	parts = append(parts, elapsed)
	if loc != "" {
		parts = append(parts, loc)
	}
	return strings.Join(filterEmpty(parts), tui.ChromeSep)
}

func filterEmpty(parts []string) []string {
	out := make([]string, 0, len(parts))
	for _, p := range parts {
		if p != "" {
			out = append(out, p)
		}
	}
	return out
}

// FormatRunListEmpty is the detail-area copy when the list has no rows.
func FormatRunListEmpty(opts RunListEmptyOpts) string {
	if opts.Unavailable {
		return "run history unavailable"
	}
	if opts.FilterActive {
		return "no matching runs"
	}
	if !opts.HasMachineRuns {
		return "no workflow has run yet"
	}
	if opts.Scope == ScopeCurrent {
		return "no runs in this worktree" + tui.ChromeSep + "Ctrl+G for All"
	}
	return "no runs"
}

// RunsFooter is the list-mode footer hint plus the scope label.
// tui.FormatListFooter renders the position once. Do not embed the position here.
func RunsFooter(scope Scope, index, total int) string {
	scopeLabel := "Current"
	if scope == ScopeAll {
		scopeLabel = "All"
	}
	_ = index
	_ = total
	return strings.Join([]string{
		"tab workflows",
		"ctrl+g " + scopeLabel,
		"enter detail",
		"esc quit",
	}, tui.ChromeSep)
}

// RunDetailFooter is the detail-mode footer hint.
func RunDetailFooter() string {
	return strings.Join([]string{"esc back", "up/down scroll"}, tui.ChromeSep)
}

// DetailLines renders a detail view into single-width lines.
func DetailLines(view DetailView, width int) []string {
	switch view.Kind {
	case "starting":
		lines := formatStartingDetail(view.Workflow, view.ID, width)
		if view.Message != "" {
			lines = append(lines, asciiGlyphs(tui.Truncate(view.Message, width)))
		}
		return lines
	case "local-failure":
		head := strings.Join([]string{"LAUNCH FAILED", view.Workflow, history.DisplayRunID(view.ID)}, tui.ChromeSep)
		return []string{tui.Truncate(head, width), asciiGlyphs(tui.Truncate(view.Message, width))}
	case "history-unavailable":
		status := "RUNNING"
		if view.Finished != "" {
			status = strings.ToUpper(view.Finished)
		}
		head := strings.Join([]string{status, "HISTORY UNAVAILABLE", view.Workflow}, tui.ChromeSep)
		lines := []string{tui.Truncate(head, width)}
		for _, line := range view.Progress {
			lines = append(lines, asciiGlyphs(tui.Truncate(line, width)))
		}
		if view.Message != "" {
			lines = append(lines, asciiGlyphs(tui.Truncate(view.Message, width)))
		}
		return lines
	default:
		lines := FormatRunDetailLines(view.Blocks, width)
		if len(view.Progress) > 0 {
			lines = append(lines, "")
			for _, line := range view.Progress {
				lines = append(lines, asciiGlyphs(tui.Truncate(line, width)))
			}
		}
		return lines
	}
}

func formatStartingDetail(workflow, id string, width int) []string {
	head := strings.Join([]string{"STARTING", workflow, history.DisplayRunID(id)}, tui.ChromeSep)
	return []string{
		tui.Truncate(head, width),
		tui.Truncate("claiming run history...", width),
	}
}

func asciiGlyphs(line string) string {
	return strings.ReplaceAll(line, "…", "...")
}

// FormatRunDetailLines renders history blocks into width-bounded lines.
func FormatRunDetailLines(blocks []history.Block, width int) []string {
	var out []string
	for _, block := range blocks {
		out = append(out, blockToLines(block, width)...)
	}
	for i, line := range out {
		out[i] = asciiGlyphs(line)
	}
	return out
}

func blockToLines(block history.Block, width int) []string {
	switch block.Kind {
	case "head":
		head := strings.Join(filterEmpty([]string{
			block.Status, block.Title, block.DisplayID, block.Elapsed,
		}), tui.ChromeSep)
		return []string{tui.Truncate(head, width)}
	case "note", "error":
		return []string{tui.Truncate(block.Text, width)}
	default:
		indent := strings.Repeat("  ", block.Depth)
		outcome := ""
		if block.Outcome != "" {
			outcome = tui.ChromeSep + block.Outcome
		}
		label := strconv.Itoa(block.Ordinal) + "/" + strconv.Itoa(block.Total)
		line := tui.Truncate(indent+label+" "+block.Label+outcome, width)
		lines := []string{line}
		if block.Explanation != "" {
			lines = append(lines, tui.Truncate(indent+"  "+block.Explanation, width))
		}
		return lines
	}
}

// ScrollDetailLines returns the visible window and clamped scroll offset.
func ScrollDetailLines(lines []string, scroll, viewport int) ([]string, int) {
	maxScroll := max(0, len(lines)-viewport)
	next := min(max(0, scroll), maxScroll)
	end := min(next+viewport, len(lines))
	return lines[next:end], next
}

// ClampDetailScroll caps scroll to the legal range for lines and viewport.
func ClampDetailScroll(lines []string, scroll, viewport int) int {
	_, next := ScrollDetailLines(lines, scroll, viewport)
	return next
}

// SelectedIndex finds the list cursor for selectedID.
func SelectedIndex(items []history.Summary, selectedID string) int {
	for i, item := range items {
		if item.ID == selectedID {
			return i
		}
	}
	return 0
}

// FormatRunSummary is the one-line runs list detail block.
func FormatRunSummary(item history.Summary) string {
	return strings.Join([]string{
		history.StatusLabel(item.Status),
		item.DisplayID,
		item.CheckoutRoot,
	}, tui.ChromeSep)
}
